### PR TITLE
Modified time on games plus reaper maybe

### LINKED ISSUE #71: Add something that reaps old games
It doesn't need to be that big really, I doubt people care to look at completed games for more than an hour or something.

### PR DESCRIPTION
**Pull request by kemcbride**
Had to go through some fixing for this one, but it seems to be working. Uh, honestly, the best way to test it is by running it. I'll update with some test output (setting INTERVAL & DURATION to small numbers so you can see the reaper run as expected)

fixes #71

### CODE REVIEW

**Comment by kemcbride** [comment]
Using INTERVAL = 10s, DURATION = 30s

```
[2020-07-03T04:43:26Z INFO  team_heist_tactics] All environment variables set
[2020-07-03T04:43:26Z INFO  team_heist_tactics::manager] Created game: test
[2020-07-03T04:43:26Z INFO  actix_server::builder] Starting 8 workers
[2020-07-03T04:43:26Z INFO  actix_server::builder] Starting "actix-web-service-0.0.0.0:19996" service on 0.0.0.0:19996
[2020-07-03T04:43:36Z INFO  team_heist_tactics] Reaper: num games: 1, num games reaped this cycle: 0
[2020-07-03T04:43:46Z INFO  team_heist_tactics] Reaper: num games: 1, num games reaped this cycle: 0
[2020-07-03T04:43:56Z INFO  team_heist_tactics] Reaper: num games: 1, num games reaped this cycle: 0
[2020-07-03T04:44:06Z INFO  team_heist_tactics] Reaper: num games: 1, num games reaped this cycle: 0
[2020-07-03T04:44:16Z INFO  team_heist_tactics] Reaper: num games: 1, num games reaped this cycle: 0
[2020-07-03T04:44:26Z INFO  team_heist_tactics] Reaper: num games: 1, num games reaped this cycle: 0
[2020-07-03T04:44:31Z INFO  team_heist_tactics::manager] Created game: DianeArbus
[2020-07-03T04:44:36Z INFO  team_heist_tactics] Reaper: num games: 2, num games reaped this cycle: 0
[2020-07-03T04:44:38Z INFO  team_heist_tactics::manager] Created game: MaryEmilyEaton
[2020-07-03T04:44:46Z INFO  team_heist_tactics] Reaper: num games: 3, num games reaped this cycle: 0
[2020-07-03T04:44:46Z INFO  team_heist_tactics::manager] Created game: NancyGrossman
[2020-07-03T04:44:56Z INFO  team_heist_tactics] Reaper: num games: 4, num games reaped this cycle: 0
[2020-07-03T04:45:06Z INFO  team_heist_tactics] Reaper: num games: 4, num games reaped this cycle: 1
[2020-07-03T04:45:16Z INFO  team_heist_tactics] Reaper: num games: 3, num games reaped this cycle: 2
[2020-07-03T04:45:21Z INFO  team_heist_tactics::manager] Created game: DianeArbus
[2020-07-03T04:45:26Z INFO  team_heist_tactics] Reaper: num games: 2, num games reaped this cycle: 0
[2020-07-03T04:45:36Z INFO  team_heist_tactics] Reaper: num games: 2, num games reaped this cycle: 0
[2020-07-03T04:45:46Z INFO  team_heist_tactics] Reaper: num games: 2, num games reaped this cycle: 0
[2020-07-03T04:45:56Z INFO  team_heist_tactics] Reaper: num games: 2, num games reaped this cycle: 1
[2020-07-03T04:46:06Z INFO  team_heist_tactics] Reaper: num games: 1, num games reaped this cycle: 0
[2020-07-03T04:46:16Z INFO  team_heist_tactics] Reaper: num games: 1, num games reaped this cycle: 0
^C[2020-07-03T04:46:18Z INFO  actix_server::builder] SIGINT received, exiting
```

File: src/manager.rs [hunk 1]
```diff
@@ -26,7 +26,7 @@ pub struct JoinOptions {
 
 #[derive(Debug)]
 pub struct GameWrapper {
-    game: Game,
+    pub game: Game,
```
Comment (banool) [inline src/manager.rs:29]: We should try and keep `game` private, it's the last bastion of private state in this dang thing, everything else can be changed by anything really. Make a method instead that gets the `game_created` time.

File: src/main.rs [hunk 2]
```diff
@@ -96,6 +105,41 @@ async fn main() -> std::io::Result<()> {
         .new_game(GameOptions {}, Some("test".to_string()))
         .unwrap();
 
+    let reaper_manager_ref = game_manager_wrapper.clone();
+
+    // Game Reaper thread
+    thread::spawn(move || loop {
+        thread::sleep(Duration::from_secs(REAP_INTERVAL));
+        let test_handle = GameHandle("test".to_string());
+        let num_games = reaper_manager_ref
```
Comment (banool) [inline src/main.rs:114]: You take and release the lock just to take it again on the next line. How about when you take the lock for `game_manager_mut`, you count the games there. Also it's just a number, you shouldn't need to clone it (bc it implements Copy).

File: src/main.rs [hunk 3]
```diff
@@ -96,6 +105,41 @@ async fn main() -> std::io::Result<()> {
         .new_game(GameOptions {}, Some("test".to_string()))
         .unwrap();
 
+    let reaper_manager_ref = game_manager_wrapper.clone();
+
+    // Game Reaper thread
+    thread::spawn(move || loop {
+        thread::sleep(Duration::from_secs(REAP_INTERVAL));
+        let test_handle = GameHandle("test".to_string());
```
Comment (banool) [inline src/main.rs:113]: The string within_handle should probs be a constant (I know it isn't now, but now we're using it twice).

File: src/main.rs [hunk 4]
```diff
@@ -96,6 +105,41 @@ async fn main() -> std::io::Result<()> {
...
+        let num_games = reaper_manager_ref
+            .game_manager
+            .read()
+            .unwrap()
+            .games
+            .len()
+            .clone();
+
+        let mut game_manager_mut = reaper_manager_ref.game_manager.write().unwrap();
+        let games = game_manager_mut.games.clone();
```
Comment (banool) [inline src/main.rs:123]: You don't really need to clone this, you could just iterate through it on line 126 and clone the `game_wrapper` when you insert it into unreaped games.

File: src/main.rs [hunk 5]
```diff
@@ -96,6 +105,41 @@ async fn main() -> std::io::Result<()> {
...
+        let mut unreaped_games: HashMap<GameHandle, std::sync::Arc<RwLock<GameWrapper>>> =
+            HashMap::new();
+        for (handle, game_wrapper) in games {
+            let creation_time = game_wrapper.read().unwrap().game.game_created;
+            if handle == test_handle || (creation_time + REAP_DURATION) > get_current_time_secs() {
+                unreaped_games.insert(handle, game_wrapper);
+            }
```
Comment (banool) [inline src/main.rs:130]: It'd be good to have an else block saying the names of the games you _are_ reaping.

File: src/main.rs [hunk 6]
```diff
@@ -96,6 +105,41 @@ async fn main() -> std::io::Result<()> {
...
+        }
+        let num_reaped_games = num_games - unreaped_games.len();
+        info!(
+            "Reaper: num games: {}, num games reaped this cycle: {}",
+            num_games, num_reaped_games,
+        );
+        // Only update if the number of games changed (ie. some were reaped)
+        if unreaped_games.len() != num_games {
+            game_manager_mut.games = unreaped_games;
+        }
```
Comment (banool) [inline src/main.rs:140]: Probs cheaper just to assign it, it's just a pointer.

File: src/main.rs [hunk 7]
```diff
@@ -96,6 +105,41 @@ async fn main() -> std::io::Result<()> {
         .new_game(GameOptions {}, Some("test".to_string()))
         .unwrap();
 
+    let reaper_manager_ref = game_manager_wrapper.clone();
+
+    // Game Reaper thread
+    thread::spawn(move || loop {
```
Comment (banool) [inline src/main.rs:111]: It'd be nice if this whole thing was its own function probably, in a different file like `periodic.rs`.

File: src/main.rs (hunk 5 above)
Comment (kemcbride) [reply src/main.rs:130]: wait, like something to log through each of the names of the reaped games? (seems legit to me)

File: src/main.rs (hunk 6 above)
Comment (kemcbride) [reply src/main.rs:140]: will check this. i think like, i added this when i was getting panics when I had the logic for reaping games backwards

File: src/main.rs (hunk 7 above)
Comment (kemcbride) [reply src/main.rs:111]: seeemsss... doable... > has not created a new file but it can't be that hard

File: src/main.rs (hunk 5 above)
Comment (banool) [reply src/main.rs:130]: Ya that.

File: src/main.rs (hunk 7 above)
Comment (banool) [reply src/main.rs:111]: Ya just a fresh file and add it to `lib.rs`.

Comment (kemcbride) [reply src/main.rs:111]: lib.rs that was the trick here (y)

**Comment by kemcbride** [comment]
Yeah ok. I do not know how to do that right now... but I'll file a task and chuck it in the "easy improvements" queue.

**Pull request merged**